# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [13]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric

In [14]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [15]:
x.shape

(506, 13)

In [16]:
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

<br><br><br>

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/BTVYXQ1.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [17]:
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples
    selecting_rows = np.random.choice(len(input_data),303,replace=False) # 60% of 506
    replacing_rows = np.random.choice(len(selecting_rows),203,replace=False) # ~40% of 506
    idx = np.random.choice(list(range(3,input_data.shape[1]+1)),1)
    selecting_columns = np.random.choice(input_data.shape[1],idx[0],replace=False) # column size must be >=3 and <13
    # for a individual row multiple column should be selected, so we put like this
    sample_data = input_data[selecting_rows[:,None],selecting_columns]
    target_of_sample_data = target_data[selecting_rows]
    
    replicated_sample_data = sample_data[replacing_rows]
    target_of_replicated_sample_data = target_of_sample_data[replacing_rows]
    
    final_sample_data = np.vstack((sample_data,replicated_sample_data))
    final_target_data = np.vstack((target_of_sample_data.reshape(-1,1),target_of_replicated_sample_data.reshape(-1,1)))

    return list(final_sample_data),list(final_target_data),list(selecting_rows),list(selecting_columns)
    #note please return as lists

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [18]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [19]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]

for i in range(0,30):
    a,b,c,d = generating_samples(x,y)
    list_input_data.append(a)
    list_output_data.append(b)
    list_selected_row.append(c)
    list_selected_columns.append(d)
    

<font color='cyan'> <b>Grader function - 2 </b></font>

In [20]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [21]:
from sklearn.tree import DecisionTreeRegressor

list_of_all_models = []
dummy_models = []
for i in range(0,30):
    dummy_models.append("clf"+str(i))
for i in range(0,30):
    dummy_models[i] = DecisionTreeRegressor(max_depth=None) #for getting more overfit model
    dummy_models[i].fit(list_input_data[i],list_output_data[i])
    list_of_all_models.append(dummy_models[i])

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [22]:
predicted_each_model_y = []
total_predicted = []

# Here we predict 506 data points in our actual data taken from boston across all 30 models
# As we are passing our original data, column sampling should be seen carefully
# So for each model, we will get 506 predicted outputs
# so totally we will have 30*506 shaped array
for i in range(0,30):
    predicted_each_model_y =[]
    for j in range(0,506):
        temp = dummy_models[i].predict(np.array(x[j][list_selected_columns[i]].reshape(1,-1)))
        predicted_each_model_y.append(temp[0])
    total_predicted.append(predicted_each_model_y)

# so in that 30*506 shaped array, 
## in the 0th row - we will have prediction for 1st data point across each model
## similarly, in the 1st row - we will have prediction for 2nd data point across each model

# so, by this understanding we can add all of them in a parallel basis
# converting into arrays for simple addition of elements in a element wise manner
# So at the end of this, we will get an 1*506 array, which will have median of predict values
# that came from all the models

pred_array = np.median(np.asarray(total_predicted),axis=0)

# Here, one key point is actual data will not change for each model, its constant through out
# so no need of calculations for actual y values
# For calculating MSC
mse = 0
for i in range(len(y)):
    mse += ((y[i]-pred_array[i])**2)  
mse/=len(y)
print("Mean Square error:",mse)

Mean Square error: 0.12433794466403164


<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [23]:
# Remember - each model is built on sample, so count of samples = count of models
final_temp = []
# listing down where does each data point not part of the sample
# at end of each i value, we get the models(samples) which that data point is not part of.
for i in range(0,506):
    temp=[]
    for j in range(0,30):
        if i not in list_selected_row[j]:
            temp.append(j)
    final_temp.append(temp)        

# Now we got the samples(models) for each datapoint which its not part of
# so we can find predictions on that models alone for each data point
oobs_final = []
for i in range(len(final_temp)):
    oobs_each = []
    for j in final_temp[i]:
        oobs_each.append(dummy_models[j].predict(np.array(x[i][list_selected_columns[j]].reshape(1,-1))))
    med = np.median(oobs_each)
    oobs_final.append(med)

# Now same like MSE calculate the OOBS score
oobs = 0
for i in range(0,506):
    oobs+=((oobs_final[i]-y[i])**2)

oobs/=len(y)
print("OOBS score:",oobs)

OOBS score: 16.553152173913045


**Observations:**<br>
*1. Here OOB score will helpful because it will give the approximation of error in a better way because when we are calculating OOB score, we are not touching the models taht includes the data point and only caring on the models where there is no data point and calculating the predicted value & then taking the median from only those boot strapped samples*<br>
*2. Where as in MSE score, we will consider all the models predicted output, so there will be chance of increasing the error for a future data point and it will be hard to trust the MSE score got from the bootstrapped samples*

# <font color='blue'><b>Task 2</b></font>

In [24]:
## Taking all the necessary modules from the previous sections

def clt_data(x,y):
    
    list_input_data =[]
    list_output_data =[]
    list_selected_row= []
    list_selected_columns=[]

    for i in range(0,30):
        a,b,c,d = generating_samples(x,y)
        list_input_data.append(a)
        list_output_data.append(b)
        list_selected_row.append(c)
        list_selected_columns.append(d)
        
    list_of_all_models = []
    dummy_models = []
    for i in range(0,30):
        dummy_models.append("clf"+str(i))
    for i in range(0,30):
        dummy_models[i] = DecisionTreeRegressor(max_depth=None) #for getting more overfit model
        dummy_models[i].fit(list_input_data[i],list_output_data[i])
        list_of_all_models.append(dummy_models[i])
    
    #### MSE Block ####
    predicted_each_model_y = []
    total_predicted = []
    for i in range(0,30):
        predicted_each_model_y =[]
        for j in range(0,506):
            temp = dummy_models[i].predict(np.array(x[j][list_selected_columns[i]].reshape(1,-1)))
            predicted_each_model_y.append(temp[0])
        total_predicted.append(predicted_each_model_y)

    pred_array = np.median(np.asarray(total_predicted),axis=0)

    mse = 0
    for i in range(len(y)):
        mse += ((y[i]-pred_array[i])**2)  
    mse/=len(y)
    
    #### OOBS Block ####
    final_temp = []
    for i in range(0,506):
        temp=[]
        for j in range(0,30):
            if i not in list_selected_row[j]:
                temp.append(j)
        final_temp.append(temp)        

    oobs_final = []
    for i in range(len(final_temp)):
        oobs_each = []
        for j in final_temp[i]:
            oobs_each.append(dummy_models[j].predict(np.array(x[i][list_selected_columns[j]].reshape(1,-1))))
        med = np.median(oobs_each)
        oobs_final.append(med)

    oobs = 0
    for i in range(0,506):
        oobs+=((oobs_final[i]-y[i])**2)

    oobs/=len(y)
    
    return mse,oobs

train_mse = []
oob_score = []
for i in range(0,35):
    mse,oobs = clt_data(x,y)
    train_mse.append(mse)
    oob_score.append(oobs)
    
print("List of Train MSE scores:",train_mse)
print("List of OOB scores:",oob_score)   

List of Train MSE scores: [0.23584609683794466, 0.13140933794466408, 0.024249011857707548, 0.12255434782608701, 0.06521638299588609, 0.05632411067193672, 0.04359189723320159, 0.07569717640009059, 0.25922357742591845, 0.0723350680720246, 0.022099802371541497, 0.2764113416470639, 0.052187835480407944, 0.09988735177865607, 0.06590415019762844, 0.08437527454479751, 0.09640143551526219, 0.03560400197628455, 0.06577294685990334, 0.11817637994351338, 0.07013833992094866, 0.12581027667984182, 0.03155632411067191, 0.10968379446640321, 0.06948342116820376, 0.04857707509881423, 0.09922430830039525, 0.018739015151515142, 0.05561758893280635, 0.06230854743083006, 0.10360464568367446, 0.06466403162055342, 0.034584053853754945, 0.08915598193577898, 0.11457894158981109]
List of OOB scores: [12.35296882411067, 13.434189185074247, 12.283552371541496, 15.173513043478257, 14.085285551434902, 14.286554545454543, 15.386329051383402, 13.308660428626961, 13.875009556383496, 15.132301987526974, 11.959471146245

In [25]:
import scipy.stats as st

mean_train = np.mean(np.asarray(train_mse))
mean_oob = np.mean(np.asarray(oob_score))

stderr_train = st.sem(np.asarray(train_mse))
stderr_oob = st.sem(np.asarray(oob_score))

# Assuming given Confidence-level is 0.95
# so what ever we consider will lie in [<mean/std>-2*(sigma/sqrt(n)),<mean/std>+2*(sigma/sqrt(n))]

train_ci_lower = mean_train-(2*stderr_train)
train_ci_upper = mean_train+(2*stderr_train)
train_ci=[train_ci_lower,train_ci_upper]
print("C.I for train_MSE:",train_ci)
oob_ci_lower = mean_oob-(2*stderr_oob)
oob_ci_upper = mean_oob+(2*stderr_oob)
oob_ci=[oob_ci_lower,oob_ci_upper]
print("C.I for OOB score:",oob_ci)

C.I for train_MSE: [0.06798186869035976, 0.10921778122532731]
C.I for OOB score: [13.12109769275497, 14.015355786957192]


**Observations:**<br>
*1. By these scores, we can say that 95% of the times, actual population train_MSE will lie in the range of (0.06798186869035976, 0.10921778122532731)*<br>
*2. where as, actual population OOB_score will lie in the range of (13.12109769275497, 14.015355786957192)*

# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [35]:
xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60]
yq = []
for i in range(0,30):
    temp = dummy_models[i].predict(np.asarray(xq)[list_selected_columns[i]].reshape(1,-1))
    yq.append(temp[0])

predicted_yq = np.median(yq)
print("The predicted output for given xq:",predicted_yq)

The predicted output for given xq: 18.5
